### ISSUE TITLE
ACL Permission for Non-Super User

### ISSUE THREAD

**Issue by amin-is** [iegomez/mosquitto-go-auth#225]
I am using Mysql Backend. Here no need any ACL Permission for a Super user it is supported Wildcards Permission for all topic. Even no need to put any topic name inside ACL table.

But for Non-Super user need this all 4 Permission individually inside Mysql table for access a topic .
I am using below mysql query in side of ACL table to given full permission for a Non-Super user.

```
insert into mosquitto_acl(mosquitto_user_id, topic, rw) values(1,'topic/abcd',1);
insert into mosquitto_acl(mosquitto_user_id, topic, rw) values(1,'topic/abcd',2);
insert into mosquitto_acl(mosquitto_user_id, topic, rw) values(1,'topic/abcd',3);
insert into mosquitto_acl(mosquitto_user_id, topic, rw) values(1,'topic/abcd',4);
```
_____
My Question is...
Is it possible to put all ACL permission using a single digit form for a Non-Super user?
An example  here readonly(1), writeonly(2), readwrite(3) and subscribe(4) total 1+2+3+4 = 10
Inside mysql table for (rw) value i will put 10 to full permission for a topic for Non-Super user.

e.g;

`insert into mosquitto_acl(mosquitto_user_id, topic, rw) values(1,'topic/abcd',10);`  [For All Permission]

`insert into mosquitto_acl(mosquitto_user_id, topic, rw) values(1,'topic/abcd',7); ` [For readwrite(3) and subscribe(4) Permission]

(Currently it is not working but i just imagine if it will work then no need to write extra lines for given ACL Permission For a Non-Super User)

Hope you understand my query about ACL.

_//English is not my native Language_

**Comment by iegomez** [comment]
I understand your concern, but it has a flaw: does `4` stand for `subscribe`, or is it `read + readwrite` (1 + 3)?. I know `read + readwrite` make no sense semantically, but then again why would you grant `1, 2, 3, 4` to a user when all you need is `3, 4` to achieve the same? Or why not simply have it be a superuser?

The thing is `3` is a legacy value from when Mosquitto didn't have `MOSQ_ACL_SUBSCRIBE` that I think was available on the famous `mosquitto-auth-plug` (can't really remember, I may have seen it elsewhere) that I took inspiration from when first writing this plugin. Mosquitto doesn't have that access value for their static ACLs actually (again, I think it was that way, at least originally, haven't seen the reworked auth work in 2.x), but since it's available in the plugin it would break backwards compatibility to remove it. It was probably a mistake to do add it, but what's done is done.

(One again, I think) Mosquitto actually follows a very desirable Unix like permissions scheme where you basically have a binary string for read, write and execute, i.e. `rwx`, that it fits nicely and unambiguously in a single octal number.

That's all to say: no, in the current scheme, you can't do that. But who knows, I may change it to follow a sane one and just bump the version, but I'm sorry to have to warn you not to hold your breath: I'm not paying much attention to the plugin lately and don't see that I'll do so in the near future either. Sorry about that.

**Comment by amin-is** [comment]
Thanks for your Reply. I am very glad to use your plugin. I think so a thing that is a Super user can access all permission without ACL. on the other hand Non-Super user need all permission. When a Non-Super user want to send, receive and subscribe a messege to a another Non-Super users topic then need to put all the permission separately that my problem. And also i feel when all user of a broker are Super user then will less security for another user too. Because a Super user are Wildcard Support power and can send, receive and subscribe message any topic to any users. If it will happen for mysql backend that a single value inside (rw) mysql table can be given full access permission for a Non-Super user to Send, Receive and Subscribe power for a another Non-Super user then it will be awesome feature.

Best Regards

**Comment by iegomez** [comment]
Yeah, I get you, but my point is that if a user is supposed to be able to do everything, then it's the same as setting it as a superuser, there's no difference. You can still have other users that don't have the superuser flag set and have explicit permissions as needed.
Also, as I mentioned, having `readwrite + subscribe` is effectively the same as having all 4 permissions if you don't want to set superusers, no need to also set `read` and `write` because that's already covered by `readwrite`.

**Comment by amin-is** [comment]
Thanks for your suggestion. I hope you will add multi level permission feature in update version. like 1+2+3+4 permission For Non-Super user as single number word. e.g; [`insert into mosquitto_acl(mosquitto_user_id, topic, rw) values(1,'topic/abcd',10);`] Because Super user can grant access wildcard '#' without put ACL list. Where Non-Super user can not access wildcard '#'  If not put into ACL list. I do not want to give wild card permission to a non-super user. i specified a topic for a user but need individually all 4 permission otherwise not working read write and subscribe for non-super user. It is my expectation that support all permission for Non-super user as a single number word inside `rw` table. If you update this feature please make it also support for MongoDB Backend.

**Comment by iegomez** [comment]
I reopened it as a reminder in case I get some time.

**Comment by amin-is** [comment]
@iegomez Thank you very much. I follow this idea and use this query and work for me everything that i exactly want.
`WHERE (username = ?) AND (rw = ? OR rw=3)`
Now i put specific permission value only 3 `rw`  for a topic and can readwrite + subscription of this topic
My problem is 100% solved i am using mosquitto version 2.0.14 on Ubuntu 20.04 & MariaDB 10

**Comment by DerEnderKeks** [comment]
First of all, thanks for this project!

I created a complete schema for PostgreSQL and found a lot of useful information in the comments here, so I though maybe some else is interested: https://gitlab.com/DerEnderKeks/mosquitto-go-auth-postgres

Btw.: `(rw | $2) > 0` would always be true, since ORing just adds up the values. (`(rw & $2) = $2` would work)

**Comment by iegomez** [comment]
> First of all, thanks for this project!
> 
> I created a complete schema for PostgreSQL and found a lot of useful information in the comments here, so I though maybe some else is interested: https://gitlab.com/DerEnderKeks/mosquitto-go-auth-postgres
> 
> Btw.: `(rw | $2) > 0` would always be true, since ORing just adds up the values. (`(rw & $2) = $2` would work)

Haha, yeah, probably a lapsus and I was looking for an AND instead, thanks for setting the record straight.

**Comment by akshayvaji** [comment]
@amin-is 

> @iegomez Thank you very much. I follow this idea and use this query and work for me everything that i exactly want. `WHERE (username = ?) AND (rw = ? OR rw=3)` Now i put specific permission value only 3 `rw` for a topic and can readwrite + subscription of this topic My problem is 100% solved i am using mosquitto version 2.0.14 on Ubuntu 20.04 & MariaDB 10

Can you please share the database value you put on, the Full database structure and auth_opt_mysql_aclquery, please

I have the same problem you faced, still not getting any idea to fix this
